### DISCUSSION TITLE
>2GB file upload component (or update FileEdit)?

### DISCUSSION THREAD

**Discussion by sturlath** [Megabit/Blazorise#3241] [General]
Any change to have a Large File Upload component for Blazor WASM (or update [FileEdit](https://blazorise.com/docs/components/file))? There seems to be a [hard file size limit for Blazor of Int32.MaxValue = 2147483647 bytes ](https://github.com/dotnet/aspnetcore/issues/33638#issuecomment-865630998).

When uploading a >2GB file I can't read it to a buffer and then send it to external API. I would either want to read it in and directly stream it to an API or brake it down into "chunks" and push it to API like e.g [Syncfusion does](https://www.syncfusion.com/blazor-components/blazor-file-upload) does.

To be sure if there is a <2GB limit I asked [this question](https://github.com/dotnet/aspnetcore/pull/33900#discussion_r771861467) but its totally unclear to me if it's still a limit in .net 6 (I will be moving there next week).

**Comment by sturlath** [comment]
Also the OpenReadStream [example](https://blazorise.com/docs/components/file/) always throws this exception with very small files (1-2MB) after hitting `OnWritten()` once

`Value was either too large or too small for an Int32`

I even found a [similar code](https://thecodebuzz.com/read-large-file-into-a-byte-array-with-chunks-csharp/) that I could not get running either... humm..

**Reply by David-Moreira** [reply]
Hello I just tested `OpenReadStream` and it does not throw the reported exception. Can you provide a repository?
![image](https://user-images.githubusercontent.com/22283161/146684764-b6049344-eedc-422b-b2fc-46b2bc8bf31c.png)

As for the large files, I'm not entirelly sure about that limitation have you tested it with Blazorise? Like this? 
@stsrki ?
```
                using ( var stream = new MemoryStream() )
                {
                    // Here we're telling the FileEdit where to write the upload result
                    await file.WriteToStreamAsync( stream );

                    // Once we reach this line it means the file is fully uploaded.
                    // In this case we're going to offset to the beginning of file
                    // so we can read it.
                    stream.Seek( 0, SeekOrigin.Begin );

                    // Use the stream reader to read the content of uploaded file,
                    // in this case we can assume it is a textual file.
                    using ( var reader = new StreamReader( stream ) )
                    {
                        fileContent = await reader.ReadToEndAsync();
                    }
                }
```

**Reply by sturlath** [reply]
Very strange... I'll create a repo for you (if that fails.. else there is something else going on)

**Comment by stsrki** [comment]
I don't remember we have any limit on the file size. We have our own implementation for the file uploading that is not using any code from the native Blazor file input. If this is a Blazor/WebAssembly limitation I also don't think we could do anything.

**Reply by sturlath** [reply]
Yes the information about limits (if any) and how to do big file uploads are very unclear and hard to read into. This has taken way to much time at my end. 

That's why I decided to just use Synfusion's File Upload component but **of course** I´m not even getting the **basics** of that to work at all with the apb.io framework, so I haven't gotten to trying out chunks! Hopefully I get that[ question](https://support.abp.io/QA/Questions/2299/IListIFormFile-is-empty-when-using-Blazor-UI-to-upload-file-but-returns-file-if-using-Swagger-UI) answered. Just wanted to update you and others down the line.

**Reply by sturlath** [reply]
The question has been [answered ](https://github.com/dotnet/aspnetcore/pull/33900#discussion_r772544607) and there **is** a <2GB limitation but only for .net 5 (and lower). 

But I'll be using Syncfusions upload component since it has lots of cool inbuilt features (progressbar,drag and drop,remove from server [etc](https://www.syncfusion.com/blazor-components/blazor-file-upload)) and if I need to upload huge files they have [chunking ](https://blazor.syncfusion.com/documentation/file-upload/chunk-upload) support also, **else I would just use the one from you guys!** 

It would be awesome to see some of these features in Blazorise one day

**Reply by stsrki** [reply]
Shouldn't be that hard to implement I think.

@David-Moreira ?

**Reply by sturlath** [reply]
@stsrki was there a ticket for this? I would love to get rid of Syncfusion and shrink the WASM size...

**Reply by stsrki** [reply]
@David-Moreira You have worked with IJSStreamReference recently. So I think this should be fixed in v1.1. Can you confirm?

**Reply by David-Moreira** [reply]
I remember testing with a few big files, but I don't remember if they were over >2GB. I need to test it. I need to open an issue so I can keep track of this. Will let you know the results.

**Reply by David-Moreira** [reply]
On another note, @sturlath since you're so keen on moving to our `FileEdit` / `FilePicker` if you have the time please test our https://github.com/Megabit/Blazorise/pull/3745 which brings improvements to `FileEdit` and the new component `FilePicker` with more inbuilt features out of the box.
See if there's any important feature we might be missing.

**Reply by sturlath** [reply]
Awesome! @David-Moreira I´ll take it for a spin in 1-3 days and give you as much feedback as I can!
